apps/prairielearn/src/pages/instructorAssessmentAccessOverrides/instructorAssessmentAccessOverrides.js

```diff
@@ -0,0 +1,193 @@
+    }
+    const group_result = await sqldb.queryZeroOrOneRowAsync(sql.select_group_in_assessment, {
+      group_name: group_name,
+      course_instance_id: course_instance_id,
+      assessment_id: assessment.id,
+    });
+    if (group_result.rows.length > 0) {
+      return { user_id: null, group_id: group_result.rows[0].id };
+    } else {
+      throw error.make(400, 'Group not found in this assessment.');
+    }
+  }
+  if (uid) {
+    if (group_name) {
+      throw error.make(400, 'Student UID is required for individual work assessments.');
+    }
+    const user = await selectUserEnrolledInCourseInstance({
+      uid: uid,
+      course_instance_id: course_instance_id,
+    });
+    if (!user) {
+      throw error.make(400, `User ${uid} is not enrolled in this course instance.`);
+    }
+
+    return { user_id: user.user_id, group_id: null };
+  } else {
+    throw error.make(400, 'Student UID or Group Name is required.');
+  }
+}
+
+async function selectUserEnrolledInCourseInstance({ uid, course_instance_id }) {
+  const user = await selectUserByUid(uid);
+  if (!user) return null;
+
+  const enrollment = await getEnrollmentForUserInCourseInstance({
+    user_id: user.user_id,
+    course_instance_id: course_instance_id,
+  });
+  if (!enrollment) return null;
+
+  return user;
+}
+
+router.get(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (!res.locals.authz_data.has_course_instance_permission_view) {
+      throw error.make(403, 'Access denied (must be a student data viewer)');
+    }
+    const result = await sqldb.queryAsync(sql.select_assessment_access_policies, {
+      assessment_id: res.locals.assessment.id,
+    });
+    res.locals.policies = result.rows;
+    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (!res.locals.authz_data.has_course_instance_permission_edit) {
+      throw error.make(403, 'Access denied (must be a student data editor)');
+    }
+
+    if (
+      req.body.__action === 'add_new_override' &&
+      res.locals.authz_data.has_course_instance_permission_edit
+    ) {
+      const { user_id, group_id } = await getUserOrGroupId({
+        course_instance_id: res.locals.course_instance.id,
+        assessment: res.locals.assessment,
+        uid: req.body.student_uid,
+        group_name: req.body.group_name,
+      });
+      await sqldb.queryAsync(sql.insert_assessment_access_policy, {
+        assessment_id: res.locals.assessment.id,
+        created_by: res.locals.authn_user.user_id,
+        credit: req.body.credit,
+        end_date: new Date(req.body.end_date),
+        group_name: req.body.group_name || null,
+        note: req.body.note || null,
+        start_date: new Date(req.body.start_date),
+        group_id: group_id || null,
+        user_id: user_id || null,
+      });
+      
+      const inserted = await insertAuditLog({
```

Comment:
I think you missed a part of my comment requesting audit logs:

> Please make sure that you wrap the action (create, update, etc) and the audit log insertion in a transaction.

You can see an example of this on another page here:

https://github.com/PrairieLearn/PrairieLearn/blob/3ba96f9c20668a79ba4e3dfb3045b8fa9805a2e6/apps/prairielearn/src/ee/pages/institutionAdminCourse/institutionAdminCourse.ts#L69

Reply:
Done in [a046c89](https://github.com/PrairieLearn/PrairieLearn/pull/9103/commits/a046c897ce189522e2d80ea862b69070929e2f65)

Reply:
Please review my comment: you need to wrap the audit log insertion *and* the corresponding access override insert/update/delete in a transaction. The transaction gives us atomicity (either all actions succeed, or none do). If this is your first exposure to that concept, Wikipedia has a decent intro: https://en.m.wikipedia.org/wiki/Atomicity_(database_systems)

Reply:
Thank you, sorry about the confusion, I will look into the wikipedia page and make these changes

Reply:
I took some time to review and made the changes in [ae288e4](https://github.com/PrairieLearn/PrairieLearn/pull/9103/commits/ae288e40b820edeb3d34626919c905f0b5aed14c)